server/asset/btc/addresser.go

```diff
@@ -0,0 +1,109 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package btc
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"decred.org/dcrdex/server/asset"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcutil/hdkeychain"
+)
+
+// AddressDeriver generates unique addresses from an extended public key.
+type AddressDeriver struct {
+	params   *chaincfg.Params
+	storeIdx func(uint32) // e.g. callback to a DB update
+	xpub     *hdkeychain.ExtendedKey
+
+	mtx  sync.Mutex
+	next uint32 // next child index to generate
+}
+
+// NewAddressDeriver creates a new AddressDeriver for the provided extended
+// public key, HDKeyIndexer, and network parameters. Note that if the source
+// wallet has accounts, the extended key should be for an account.
+func NewAddressDeriver(xpub string, keyIndexer asset.HDKeyIndexer, chainParams *chaincfg.Params) (*AddressDeriver, error) {
```

Comment:
I was actually making a point _not_ to have the `AddressDeriver` be a `Backend` method because there's no need to have a connection to a node to derive addresses.  Setting up a backend if you just want an `AddressDeriver` would be an unnecessary burden.

In fact, I was considering going ever further and separating it from the `Driver`, but it would be more code to create them by asset symbol.  Hence this convenience method: https://github.com/decred/dcrdex/pull/1202/commits/528b73953157edbd828937e3d190d8cccf1cf898#diff-51dc1aa092977ea75e0a2e2e2bc5263254edd21ca3f653b0a27035c5c667c829R62

If I were writing some other package or app that needed to derive Bitcoin addresses specifically, I'd definitely just import the `server/asset/btc` package and use `NewAddressDeriver`.

I completely acknowledge that the `fmt.Printf`s are sucky though.  They were mostly there for debugging, so I wouldn't miss them if they were just removed.

Does that makes sense?  I would find it really unfortunate to make every type and function in the server/asset/{symb} packages require a configured Backend to be useable.

Reply:
I think you'll find commit https://github.com/decred/dcrdex/pull/1202/commits/e8f9e57382b7555a22b071a50894ae02f013ee91 more palatable.

Instead of being lazy and type-asserting the `Backend` to see if it is also `NewAddresser`, I've updated the `Driver` implementations so the addressers can be generated by asset symbol just like the `Backend`s.  This way no `Backend` is required to use a `NewAddresser` (it doesn't need a node), and one can be created either via (1) `asset.NewAddresser(symbol, ...` **or** (2) directly if you have the specific asset's package in mind like `server/asset/dcr.NewAddresser`